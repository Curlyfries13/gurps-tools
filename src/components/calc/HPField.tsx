import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import {
  updateDisplayHP,
  updateDisplayCurrHP,
} from 'src/redux/actions/characterActions';
import { RootState } from 'src/redux';

const HPField = ({
  hp,
  currHP,
  updateDisplayHP,
  updateDisplayCurrHP,
}: Props) => {
  function handleUpdateHP(event: React.ChangeEvent<HTMLInputElement>): void {
    const { value } = event.target;
    updateDisplayHP(value);
  }

  function handleUpdateCurrentHP(
    event: React.ChangeEvent<HTMLInputElement>
  ): void {
    const { value } = event.target;
    updateDisplayCurrHP(value);
  }

  // there should only be one of these components: all id's are designated as
  // if this is the case
  return (
    <>
      <label
        htmlFor='max-hp'
        className='col col-form-label d-flex justify-content-center'
      >
        HP
      </label>
      <div className='d-flex'>
        <div>
          <input
            type='text'
            aria-label='current hit points'
            className='form-control'
            id='current-hp'
            value={currHP}
            onChange={handleUpdateCurrentHP}
          />
        </div>
        <p className='fs-4 mx-2 text-muted'>/</p>
        <div>
          <input
            type='text'
            aria-label='total hit points'
            className='form-control'
            id='max-hp'
            value={hp}
            onChange={handleUpdateHP}
          />
        </div>
      </div>
    </>
  );
};

function mapStateToProps(state: RootState) {
  return {
    hp: state.displayHP,
    currHP: state.displayCurrHP,
  };
}

const mapDispatchToProps = {
  updateDisplayHP,
  updateDisplayCurrHP,
};

const connector = connect(mapStateToProps, mapDispatchToProps);

type Props = ConnectedProps<typeof connector>;

export default connector(HPField);
