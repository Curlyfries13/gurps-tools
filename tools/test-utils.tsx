// Using code from https://redux.js.org/recipes/writing-tests#connected-components
import React from 'react';

// test-utils.js
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
// Import your own reducer
//
// console.log(React.createElement);

import * as reducer from 'src/redux/reducers';


const customRender = (
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) => {
  const Wrapper =({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
