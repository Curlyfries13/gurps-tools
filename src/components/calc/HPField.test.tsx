import React from 'react';

import { RenderResult } from '@testing-library/react';
import HPField from './HPField';

import { render, fireEvent, waitFor, screen } from 'tools/test-utils';
import store from 'src/redux';

const mockState = {
  hp: 0,
  currHP: 0,
};

it('updates the current HP to the total HP ', async () => {
  render(<HPField />, {
    initialState: mockState,
    store: store,
  });
  fireEvent.change(screen.getByLabelText(/total hit points/i), {
    target: { value: '10' },
  });
  const test = screen.getByLabelText(/current hit points/i);
  expect(test.value).toBe('10');
});
