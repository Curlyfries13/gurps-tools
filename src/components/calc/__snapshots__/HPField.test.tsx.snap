// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="col col-form-label d-flex justify-content-center"
        for="max-hp"
      >
        HP
      </label>
      <div
        class="d-flex"
      >
        <div>
          <input
            aria-label="current hit points"
            class="form-control"
            id="current-hp"
            type="text"
            value=""
          />
        </div>
        <p
          class="fs-4 mx-2 text-muted"
        >
          /
        </p>
        <div>
          <input
            aria-label="total hit points"
            class="form-control"
            id="max-hp"
            type="text"
            value="0"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <label
      class="col col-form-label d-flex justify-content-center"
      for="max-hp"
    >
      HP
    </label>
    <div
      class="d-flex"
    >
      <div>
        <input
          aria-label="current hit points"
          class="form-control"
          id="current-hp"
          type="text"
          value=""
        />
      </div>
      <p
        class="fs-4 mx-2 text-muted"
      >
        /
      </p>
      <div>
        <input
          aria-label="total hit points"
          class="form-control"
          id="max-hp"
          type="text"
          value="0"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
